### PR TITLE
Update sidebar

### PR DESCRIPTION
**Pull request by KTibow**
## Proposed change

Updates the sidebar to
- use raw links instead of paper-listbox, fixes some ugly apperance stuff and ctrl+clicking
- use m3 styling
- refactor so it isn't a megacomponent

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
NA
## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by KTibow** [comment]
| Original | M3 Example | Implementation |
|--------|--------|--------|
| image | image | image |
| image | image | image |

File: src/components/ha-sidebar.ts [hunk 1]
```diff
@@ -198,934 +41,88 @@ class HaSidebar extends SubscribeMixin(LitElement) {
...
+  private _panelHover(ev: CustomEvent) {
+    this._mouseOverItem({ target: ev.detail });
   }
 
-  private _listboxFocusIn(ev) {
-    if (this.alwaysExpand || ev.target.nodeName !== "A") {
+  private _mouseOverItem(ev: { target: HTMLElement }) {
+    const tooltipRoot =
+      this.shadowRoot?.querySelector("ha-sidebar-tooltip")?.shadowRoot;
+    const tooltip = tooltipRoot?.firstElementChild as HTMLElement;
```
Comment (bramkragten) [inline src/components/ha-sidebar.ts:101]: Why not just use a `slot`?

**Comment by bramkragten** [comment]
> use raw links instead of paper-listbox, fixes some ugly apperance stuff and ctrl+clicking

With the list element removed, we now no longer have keyboard navigation, like typeahead and simple arrow handling. Also is it now missing a lot of a11y feats.

File: src/components/ha-sidebar.ts (hunk 1 above)
Comment (KTibow) [reply src/components/ha-sidebar.ts:101]: i don't think a slot would make things easier from what i understand, the tooltip would still be in a shadow root, and you would require even more logic to traverse the slot

**Comment by KTibow** [comment]
re: a11y, after checking on my machine:
- i do need to add `tabindex="-1"` to one of the items and do that `role="option"` thing
- what is "typeahead and simple arrow handling"?
- keyboard navigation still works with shift-tab/tab and enter

**Comment by LasseRosenow** [comment]
Maybe better to just wait for the [@material/web](https://github.com/material-components/material-web) drawer component?

**Comment by bramkragten** [comment]
> re: a11y, after checking on my machine:
> 
> * what is "typeahead and simple arrow handling"?
> * keyboard navigation still works with shift-tab/tab and enter

Press a letter when focus is on the old sidebar list, it will focus the item starting with the letter you pressed

**Comment by KTibow** [comment]
@LasseRosenow there is a nav drawer component but that just wraps whatever content you want to put inside of it - could be they'll add more later as it's marked as wip
@bramkragten what a11y specs are you referring to

**Comment by bramkragten** [comment]
Basic a11y features for lists <https://w3c.github.io/aria/#list>

**Comment by KTibow** [comment]
Well I'm not sure how to move forward on this, as using paper-listbox would reintroduce the bugs like ctrl+clicking behavior. Should I make a custom implementation to handle those keys?

**Comment by bramkragten** [comment]
> Well I'm not sure how to move forward on this, as using paper-listbox would reintroduce the bugs like ctrl+clicking behavior. Should I make a custom implementation to handle those keys?

You can use `mwc-list` but that will not work with anchor links, so open in new tab etc will no longer work.

**Comment by KTibow** [comment]
I believe this PR should be ready for comment now. In my testing I didn't find any bugs, and everything from the ARIA listbox guidelines (if it was in there previously) is now working.

**Comment by silamon** [comment]
Too bad this didn't got a review yet, I've had a go and tested your changes. I didn't look at the code yet. 

The sidebar is so much smoother by removing that paper-listbox. It feels like the dashboard also loads faster. 
I hope you don't mind if I share some things that didn't work for me: 
* Reordering the dashboards: I can go in edit mode, but dragging dashboards to other places doesn't seem to move it. 
* The scrollbar takes a lot of place as can be seen in the screenshot
image

**Comment by KTibow** [comment]
i'll see if i can reproduce that

File: src/components/ha-sidebar-panels.ts [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+  return [beforeSpacer, afterSpacer];
+};
+const computePanels = memoizeOne(_computePanels);
+
+@customElement("ha-sidebar-panels")
+class HaSidebarPanels extends LitElement {
+  static shadowRootOptions = {
+    ...LitElement.shadowRootOptions,
+    delegatesFocus: true,
+  };
```
Comment (silamon) [inline src/components/ha-sidebar-panels.ts:141]: This is the reason the ordering of dashboards no longer work.

Comment (KTibow) [reply src/components/ha-sidebar-panels.ts:141]: Can't reproduce the issue with reordering dashboards
Can you send a screen recording and other relevant info

**Comment by bramkragten** [comment]
Major UI changes like these will have to be approved by our design department, CC @matthiasdebaat

File: src/components/ha-sidebar-panels.ts (hunk 2 above)
Comment (silamon) [reply src/components/ha-sidebar-panels.ts:141]: So, I can reproduce this on both Microsoft Edge and Google Chrome latest version. On Mozilla Firefox it works good. I'll see if I can create a screen recording.

**Comment by KTibow** [comment]
I see that there was a "needs design preview" label added. Assuming that refers to https://design.home-assistant.io/, I have some questions about how to implement one:
- How exactly do I display the sidebar? Am I supposed to display the whole sidebar or individual panels? If I need to display the whole sidebar:
  - What kinds of states do I need to display? Do I need to show what it looks like with devtools, on mobile, etc? If so, should I just display all the different kinds of sidebars, or provide switches that can be used?
  - Where can I learn about how to set up a fake hass?
- As someone who isn't as familiar with the frontend as someone in the core team might be, is there a specific contributing guide for how to create a design preview?
- Also, is there any specific reason a design preview was requested? The previous sidebar isn't there. Is this a norm for new UI components? Does this make it easier to review? As you can tell I'm not sure if I know enough to create a design preview.

**Comment by matthiasdebaat** [comment]
Thank you for helping us update our components to Material 3. From a UX point of view, this PR should only update the visual design, with a few exceptions:
- M3 design language is bolder than 2, which takes up more space. The disadvantage is that fewer menu items are displayed if we follow these guidelines. Our users have a lot of menu items in the sidebar, showing less would not help our users. In this case, we need to keep the height of the menu items the same as we currently have.
- While I like the M3 approach of showing a number instead of a badge, we need to keep the collapsed and expanded menu consistent. Let's keep our badge in both states.

Curious how the edit "wiggle" mode looks. Great if we can add the sidebar to our design website.

> I see that there was a "needs design preview" label added. Assuming that refers to https://design.home-assistant.io/, I have some questions about how to implement one:

@bramkragten can help you with this.

**Comment by KTibow** [comment]
> In this case, we need to keep the height of the menu items the same as we currently have.

Crazy idea but what do you think about allowing the menu items to compress if there's too many? Of course if there's too many they would just max out at the previous UI height and scroll the menu.

> Let's keep our badge in both states.

[TODO: check if the badge still breaks at 10+ numbers]

> Curious how the edit "wiggle" mode looks

Edit works as usual, except I didn't implement the wiggling yet.


To summarize, soon I'll implement
- some sort of change to the item height (once I hear what other people think about a dynamic item height)
- using the badge in more cases
- wiggling animation
- larger touch targets
- fixing merge conflicts

**Comment by matthiasdebaat** [comment]
> Crazy idea but what do you think about allowing the menu items to compress if there's too many? Of course if there's too many they would just max out at the previous UI height and scroll the menu.

Interesting idea, have to see it if we should implement it. Couple of downsides is that it can get bouncy or have unexpected behaviour for our users.

**Comment by karwosts** [comment]
`needs design preview` is supposed to trigger a github task to automatically build a new copy of design.home-assistant.io with the changes in this PR. However I believe the action can not work for PRs that originate from forks, so it won't work for this case.

**Comment by KTibow** [comment]
Looks like the "deploy preview" workflow is asking for the user to open a browser to sign in, so it failed to deploy. That's also why it took so long - it timed out.

**Comment by KTibow** [comment]
Hey, I'd just like to recap that with `3f1fe22` and `586e12f`, I added/fixed these things:
- A reduced item height
- Using the full badge when expanded
- A wiggling animation when in edit mode
- Full touch targets

If the code looks all good, this PR is ready to merge.

**Comment by KTibow** [comment]
Should be resolved on the latest commit

File: src/components/ha-sidebar-panels.ts (hunk 2 above)
Comment (KTibow) [reply src/components/ha-sidebar-panels.ts:139]: Can't reproduce w/ Chromium 117.0.5938.132 or Chrome Mobile 118.0.5993.80
Does it work if you clear your localstorage by chance

Comment (silamon) [reply src/components/ha-sidebar-panels.ts:139]: Too bad, removing localstorage doesn't solve it. It works if the shadowRootOptions are removed or delegatesFocus is set to false. Let's wait until somebody other reviews.

**Comment by silamon** [comment]
@bramkragten can this get a review somehow or is there anything that still needs to be done before review?

**Comment by bramkragten** [comment]
I don't like the text color and especially the icon color of the sidebar, is there a reason these are now so dark?
Also the text "Home Assistant" on top of the expanded sidebar looks very small?
I also miss things like the ripple effect in your implementation, but that might be ok.

**Comment by KTibow** [comment]
> I don't like the text color and especially the icon color of the sidebar, is there a reason these are now so dark?

In the dark theme or light theme? Thing is that M3 calls for `on-surface-variant`, but we only have `primary-text-color` and `secondary-text-color` (a much less visible color). While I could decrease the opacity (to 90%) to match `on-surface-variant`, since I use rgba and var in my code and you can't multiply rgba, this wouldn't work.

> Also the text "Home Assistant" on top of the expanded sidebar looks very small?

Yup, this is in compliance with Material 3. The title should be `title-small`, which is 14pt. Technically speaking it should be even smaller as Home Assistant uses a text size that is 7/8 the normal size.

> I also miss things like the ripple effect in your implementation, but that might be ok.

The previous implementation didn't have ripples, and I'm not exactly sure which ripples to use.

File: src/components/ha-sidebar-panel-config.ts [hunk 3]
```diff
@@ -30,17 +31,19 @@ class HaSidebarPanelConfig extends LitElement {
...
   @state() private _issuesCount = 0;
 
-  static styles = [haStyleSidebarItem, styles];
-
   protected render() {
     const notices = this._updatesCount + this._issuesCount;
+    const keydown = createKeydown((e) =>
+      (e.currentTarget as HTMLElement).click()
+    );
+    const keyup = createKeyup((e) => (e.currentTarget as HTMLElement).click());
```
Comment (bramkragten) [inline src/components/ha-sidebar-panel-config.ts:39]: This is still the same, as every time on render a new function is created and a new event listener is added. This should just be a class function.

**Comment by codyc1515** [comment]
Current or can be closed?

**Comment by KTibow** [comment]
Let me fix the conflicts (yet again..)

**Comment by KTibow** [comment]
Actually there have been too many changes for me to want to bother with fixing this yet again. I'm just sad that this never got a chance to get merged.

**Pull request closed without merging**